### PR TITLE
Implement optional rounding

### PR DESCRIPTION
**Pull request by ChristianZimpelmann**
### What problem do you want to solve?

Implement the possibility for optional rounding as specified in [GEP-05](https://gettsim--324.org.readthedocs.build/en/324/geps/gep-05.html). 

And implement optional rounding for one examplary function (`eink_st_tu`)

Addresses #88.

#316 was started earlier and will implement rounding for all Grundrente and Grundsicherung im Alter variables after this PR is merged.

### Todo

- [x] Implement decorators functions
- [x] Implement way to set globally whether outputs should be rounded or not (where applicable)
- [x] Add rounding parameter and decorator for `eink_st_tu`
- [x] Write tests for rounding decorator

### Open ToDos after PR

- Reintroduce rounding where applicable

### CODE REVIEW

File: gettsim/tests/test_rounding.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    calc_result = compute_taxes_and_transfers(
+        data=data, functions=[test_func], params=rounding_specs, targets=["test_func"],
+    )
+    np.array_equal(calc_result["test_func"].values, np.array(exp_output))
+
+
+@pytest.mark.parametrize(
+    "base, direction, input_values, exp_output",
```
Comment (hmgaudecker) [inline gettsim/tests/test_rounding.py:110]:
```suggestion
    "base, direction, input_values_exp_output, _ignore",
```

Then re-use the parametrization from `test_rounding`.

File: gettsim/tests/test_rounding.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    [
+        (5, "up", [100.24, 100.78], [100.24, 100.78]),
+        (0.1, "down", [100.24, 100.78], [100.24, 100.78]),
+        (0.001, "nearest", [100.24, 100.78], [100.24, 100.78]),
+    ],
+)
+def test_no_rounding(base, direction, input_values, exp_output):
+    """Check if no rounding if disabled"""
```
Comment (hmgaudecker) [inline gettsim/tests/test_rounding.py:118]:
```suggestion

```
(obvious)

File: gettsim/tests/test_rounding.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    # Define function that should be rounded
+    @add_rounding_spec(params_key="params_key_test")
+    def test_func(income):
+        return income
+
+    data = pd.DataFrame([{"p_id": 1}, {"p_id": 2}])
+    data["income"] = input_values
```
Comment (hmgaudecker) [inline gettsim/tests/test_rounding.py:126]:
```suggestion
    data["income"] = input_values_exp_output
```

File: gettsim/tests/test_rounding.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    rounding_specs = {
+        "params_key_test": {
+            "rounding": {"test_func": {"base": base, "direction": direction}}
+        }
+    }
+
+    calc_result = compute_taxes_and_transfers(
+        data=data, functions=[test_func], params=rounding_specs, targets=["test_func"],
+    )
+    np.array_equal(calc_result["test_func"].values, np.array(exp_output))
```
Comment (hmgaudecker) [inline gettsim/tests/test_rounding.py:136]:
```suggestion
    np.array_equal(calc_result["test_func"].values, np.array(input_values_exp_output))
```

**Review by hmgaudecker — COMMENTED** [review]
Looks great, thanks! Just a number of small things!

File: gettsim/tests/test_rounding.py (hunk 1 above)
Comment (ChristianZimpelmann) [reply gettsim/tests/test_rounding.py:109]: What would be the function of `_ignore`? If I understand your comment correctly, I could re-use the parametrization without an additional keyword.

Comment (ChristianZimpelmann) [reply gettsim/tests/test_rounding.py:109]: Ah, got it now :)

File: gettsim/interface.py [hunk 5]
```diff
@@ -545,11 +717,15 @@ def _partial_parameters_to_functions(functions, params):
...
         }
+        if partial_params:
+            partial_func = functools.partial(function, **partial_params)
 
-        partialed_functions[name] = (
-            functools.partial(function, **partial_params)
-            if partial_params
-            else function
-        )
+            # Make sure that rounding parameter attribute is kept
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/interface.py:723]: Changed the comment to:

```
            # Make sure that rounding parameter attribute is transferred to partial function.
            # Otherwise, this information would get lost.
            
            ```
            
            Clearer?

File: gettsim/interface.py [hunk 6]
```diff
@@ -519,6 +586,111 @@ def _root_nodes(dag):
...
+                    raise KeyError(
+                        "Both 'base' and 'direction' are expected as rounding "
+                        "parameters in the parameter dictionary. At least one of them "
+                        f"is missing at ['{params_key}']['rounding']['{task}']."
+                    )
+
+                # Add rounding
+                base = rounding_spec["base"]
+                direction = rounding_spec["direction"]
+                dag.nodes[task]["function"] = _add_rounding_to_one_function(
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/interface.py:688]: It fails within `_add_rounding_to_one_function`. I would prefer that as is it is easier to adjust when we add more potential values for `direction`.

**Comment by ChristianZimpelmann** [comment]
Great comments! Thanks!

I addressed almost all of the comments, but left a few open to give more information or to disagree with the proposed changes (in two cases). Can change it if I couldn't convince you. :)

File: gettsim/interface.py [hunk 7]
```diff
@@ -65,13 +69,87 @@ def compute_taxes_and_transfers(
...
+
+    if not debug:
+        results = results[targets]
+
+    results = _reorder_columns(results)
+
+    return results
+
+
+def set_up_dag(
```
Comment (hmgaudecker) [inline gettsim/interface.py:120]: Only realised this just now: It is pretty confusing to have one function `set_up_dag`, which calls `create_dag`. Can we do better? No time for brainstorming right now, sorry.

File: gettsim/interface.py [hunk 8]
```diff
@@ -65,13 +69,87 @@ def compute_taxes_and_transfers(
...
+    data,
```
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply gettsim/interface.py:121]: But almost trivially, no? Go through all nodes of the DAG, check if it is a function or a data column, do the check if the latter. 

Given that the function is new, I would go that far in case I am not missing anything.

[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/interface.py:121]: Usual data columns are trivial. The problem is data columns that ovewrite functions. For those, the expected data type, depends on the return type of the function, but the function is not part of the DAG and doesn't exist outside of the function as it is implemented right now.

[the remark this answers is not in the document]
Comment (hmgaudecker) [reply gettsim/interface.py:121]: Okay, just so that I understand: we would need to update the dict of expected data types defined for the input columns (can't look up right now) while looking at columns overriding functions?

What about passing (example, untested)

`{col: dtype for col, dtype in data.dtypes}`

for the moment?

I'd just like to signal that it has nothing to do with the data, really.

[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/interface.py]: I now restructured `compute_taxes_and_transfers` which now calls (among other) two functions:

- ``check_data_check_functions_and_merge_functions``
- ``prepare_functions_and_set_up_dag``

``prepare_functions_and_set_up_dag`` no longer requires ``data`` as argument. ``check_data_check_functions_and_merge_functions`` still does so (among others because `check_if_series_has_internal_type` requires the pd.series and not just the dtypes). 

But this structure should make clear that creating the DAG has nothing to do with data as was your concern if I understand correctly.

File: gettsim/interface.py [hunk 9]
```diff
@@ -65,55 +77,182 @@ def compute_taxes_and_transfers(
...
+        columns_overriding_functions=columns_overriding_functions,
+        check_minimal_specification=check_minimal_specification,
+        rounding=rounding,
+    )
+
+    # Do some checks.
+    _fail_if_root_nodes_are_missing(dag, data)
+    _fail_if_more_than_necessary_data_is_passed(dag, data, check_minimal_specification)
+
+    # We delay the data preparation as long as possible such that other checks can fail
```
Comment (hmgaudecker) [inline gettsim/interface.py:111]: Comment does not seem quite up-to-date? We may want to do the `_process_data` step earlier with other checks? Looks a bit like it should be the first one, really.

E.g., `data_cols = list(data.columns)` in `check_data_check_functions_and_merge_functions` now implicitly assumes that the input data is a DataFrame.

File: gettsim/interface.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply gettsim/interface.py:770]: Cool. I just removed creating variables just to pass them to the function, then it makes sense to my weird mind, too :wink:

File: gettsim/parameters/eink_st.yaml [hunk 10]
```diff
@@ -239,3 +239,12 @@ ertragsanteil:
       upper_threshold: inf
       rate_linear: 0.0
       intercept_at_lower_threshold: 1
+
+rounding:
```
Comment (hmgaudecker) [inline gettsim/parameters/eink_st.yaml:243]: Oh, so you did add some rounding? Sorry I missed that in the previous round. Be precise then in changes.rst

File: gettsim/policy_environment.py [hunk 11]
```diff
@@ -358,9 +361,52 @@ def _load_parameter_group_from_yaml(date, group, parameters=None):
...
+
+    Returns:
+        dictionary: Rounding parameters for the specified date
+    """
+    out = {}
+    rounding_parameters = ["direction", "base"]
+
+    # Load values of all parameters at the specified date
+    for function_name in rounding_spec.keys():
+        dates = sorted(
```
Comment (hmgaudecker) [inline gettsim/policy_environment.py:391]:
```suggestion
        policy_dates = sorted(
```

File: gettsim/policy_environment.py [hunk 12]
```diff
@@ -358,9 +361,52 @@ def _load_parameter_group_from_yaml(date, group, parameters=None):
...
+            key
+            for key in rounding_spec[function_name].keys()
+            if isinstance(key, datetime.date)
+        )
+        past_policies = [x for x in dates if x <= date]
```
Comment (hmgaudecker) [inline gettsim/policy_environment.py:396]: I find the logic (too) hard to follow here. Why not get `policy_in_place` in one step? Also, the name `past_policies` suggests that none of them is relevant anymore, maybe that's what is confusing me.

File: gettsim/shared.py [hunk 13]
```diff
@@ -57,3 +57,27 @@ def get_names_of_arguments_without_defaults(function):
             argument_names_without_defaults.append(parameter)
 
     return argument_names_without_defaults
+
+
+def add_rounding_spec(params_key):
+    """Decorator to add the parameter key where rounding specifications are
```
Comment (hmgaudecker) [inline gettsim/shared.py:63]:
```suggestion
    """Decorator adding the location of the rounding specification to a function.
```

File: gettsim/shared.py [hunk 14]
```diff
@@ -57,3 +57,27 @@ def get_names_of_arguments_without_defaults(function):
...
+    found to a function.
```
Comment (hmgaudecker) [inline gettsim/shared.py:64]:
```suggestion
```

File: gettsim/tests/test_interface.py [hunk 15]
```diff
@@ -155,3 +157,38 @@ def test_consecutive_internal_test_runs():
 
     with pytest.warns(UserWarning, match="Repeated execution of the test suite"):
         test("--collect-only")
+
+
+def test_partial_parameters_to_functions():
```
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply gettsim/tests/test_interface.py:162]: I guess you could just define it outside of the tests to avoid writing it twice? Of course you could not call it `test_func` then :smile:

File: gettsim/policy_environment.py (hunk 12 above)
Comment (hmgaudecker) [inline gettsim/policy_environment.py:396]:
```suggestion
        past_policies = [d for d in policy_dates if d <= date]
```

**Review by hmgaudecker — APPROVED** [review]
Left open a couple of small things, but basically done! Thanks!!!

File: gettsim/interface.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply gettsim/interface.py]: Cool. Just a small follow-up re the expected data left now.

File: gettsim/tests/test_interface.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply gettsim/tests/test_interface.py:162]: good point!

File: gettsim/policy_environment.py (hunk 12 above)
Comment (ChristianZimpelmann) [reply gettsim/policy_environment.py:396]: Good point about the naming! I improved naming of variables and Documentation.

> Why not get policy_in_place in one step?

To handle cases in which `policy_dates_before_date` is empty. Then taking the maximum leads to an error. (Could be also handled with try except, but I like the current version a bit more.

Alternative would be to fail already here if this happens, but I wouldn't since the user can still change `policy_params` themselves.

File: gettsim/policy_environment.py [hunk 16]
```diff
@@ -358,9 +361,55 @@ def _load_parameter_group_from_yaml(date, group, parameters=None):
...
+        # Find all specified policy dates before date.
+        policy_dates_before_date = sorted(
+            key
+            for key in rounding_spec[function_name].keys()
+            if isinstance(key, datetime.date) and key <= date
+        )
+
+        # If any rounding specs are defined for a date before the specified
+        # date, copy them to params dictionary.
+        # If not, no key for this function name is created (this will
```
Comment (hmgaudecker) [inline gettsim/policy_environment.py:401]: I don't understand the second part ("If not, ...")

File: gettsim/policy_environment.py [hunk 17]
```diff
@@ -358,9 +361,55 @@ def _load_parameter_group_from_yaml(date, group, parameters=None):
...
+        for several functions and several dates
+
+    Returns:
+        dictionary: Rounding parameters for the specified date
+    """
+    out = {}
+    rounding_parameters = ["direction", "base"]
+
+    # Load values of all parameters at the specified date.
+    for function_name in rounding_spec.keys():
```
Comment (hmgaudecker) [inline gettsim/policy_environment.py:390]: Iterate over `.items()` since we are accessing `rounding_spec[function_name]` a couple of times. 

Maybe:

```suggestion
    for function_name, this_rounding_spec in rounding_spec.items():
```
?

File: gettsim/policy_environment.py [hunk 18]
```diff
@@ -358,9 +361,55 @@ def _load_parameter_group_from_yaml(date, group, parameters=None):
...
+    Parameters
+    ----------
+    date : datetime.date
+        The date for which the policy system is set up.
+    rounding_spec : dictionary
+        The dictionary that contains rounding parameters
+        for several functions and several dates
+
+    Returns:
+        dictionary: Rounding parameters for the specified date
```
Comment (hmgaudecker) [inline gettsim/policy_environment.py:384]:
```suggestion
        dictionary: 
          - Keys: Functions to be rounded.
          - Values: Rounding parameters for the specified date
```

File: gettsim/policy_environment.py [hunk 19]
```diff
@@ -358,9 +361,55 @@ def _load_parameter_group_from_yaml(date, group, parameters=None):
...
 
+def _load_rounding_parameters(date, rounding_spec):
+    """Load rounding parameters for a specific date from a dictionary.
+
+    Parameters
+    ----------
+    date : datetime.date
+        The date for which the policy system is set up.
+    rounding_spec : dictionary
+        The dictionary that contains rounding parameters
```
Comment (hmgaudecker) [inline gettsim/policy_environment.py:380]:
```suggestion
          - Keys: Functions to be rounded.
          - Values: Rounding parameters for all dates
```

File: gettsim/policy_environment.py [hunk 20]
```diff
@@ -358,9 +361,55 @@ def _load_parameter_group_from_yaml(date, group, parameters=None):
...
+        for several functions and several dates
```
Comment (hmgaudecker) [inline gettsim/policy_environment.py:381]:
```suggestion
```

**Review by hmgaudecker — APPROVED** [review]
Nice, just documentation left, essentially.

File: gettsim/policy_environment.py (hunk 16 above)
Comment (ChristianZimpelmann) [reply gettsim/policy_environment.py:401]:
```

        # If no appropriate rounding specs are found for the requested date, 
        # no key for this function name is created (this will raise an error 
        # later if the user doesn't adjust params df before calling 
        # `compute_taxes_and_transfers`).

Comment (hmgaudecker) [reply gettsim/policy_environment.py:401]:
```
        # If no appropriate rounding specs are found for the requested date, 
        # the function will not appear in the returned dictionary. 
        # Note this will raise an error later unless the user adds an appropriate
        # rounding specification to the parameters... 
```

*(not sure about the wording in the very last bit, but we should be precise there. I don't think there is a "params df", for example?)*

Comment (ChristianZimpelmann) [reply gettsim/policy_environment.py:401]: Sounds good: `... to the parameters dictionary.`
